crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use anyhow::{ensure, Context, Result};
+use clap::ValueEnum;
+
+#[derive(clap::ValueEnum, Clone, Debug)]
+pub enum ChangelogReleaseType {
+    Major,
+    Minor,
+    Patch,
+}
+
+/// Changelog file data structure.
+#[derive(Debug, Default, PartialEq)]
+struct Changelog {
+    releases: Vec<ChangelogRelease>,
+
+    increment_count: u32,
+}
+
+impl Changelog {
+    fn push_release(&mut self, release: ChangelogRelease) {
+        self.releases.push(release);
```

Comment:
Maybe we should check here that:
- releases are in decreasing precedence order (i.e. push on empty should be 0.1.0, otherwise should be higher precedence that the top)
- can't push on top of a pre-release (i.e. if there's a pre-release it's the top)